README
$Id: README 1 2006-08-16 05:19:58Z miyagawa $

This is XML::Atom, an implementation of the API and feed format.

PREREQUISITES

    * XML::LibXML
    * LWP
    * Digest::SHA1
    * MIME::Base64
    * HTML::Parser

INSTALLATION

XML::Atom installation is straightforward. If your CPAN shell
is set up, you should just be able to do

    % perl -MCPAN -e 'install XML::Atom'

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

Benjamin Trott / cpan@stupidfool.org


Changes
$Id: Changes 110 2009-01-07 02:01:47Z miyagawa $

Revision history for XML::Atom

0.33  2008.01.06
    * Client: Send WSSE auth header only when username is set
      (Thanks to David Bourget at http://rt.cpan.org/Public/Bug/Display.html?id=42201)

0.32  2008.11.23
    * Added base and lang to Entry and Feed object. Fixed xml:base attributes.
      (Simon Wistow)

0.31  2008.11.13
    * Update Content-Type in XML::Atom::Client when the entity's version >= 1.0
      (Thanks to David Steinbrunner RT 39801)

0.30  2008.11.12
    * hopefully fix a bug where xml:base returns an empty string
      e.g. http://www.nntp.perl.org/group/perl.cpan.testers/2008/11/msg2595696.html
      (Thanks to tokuhirom http://d.hatena.ne.jp/tokuhirom/20081110/1226280757)

0.29  2008.10.25
    * Skips Unicode tests since it doesn't pass with some libxml versions (and it's not actually a bug)

0.28  2007.11.06
    * Fixed Namespace handling in extensions so that both URL and NS object work
      (Thanks to Brian Cassidy)

0.27_01 2007.10.04
    * Removes most of hacks to deal with LibXML insane unicode stuff which are fixed with 1.64

0.27 2007.09.15
    * Fixed $feed->as_xml_utf8 to work with latest XML::LibXML

0.26 2007.09.15
    * This be 0.26

0.25_02  2007.06.20
    * Fixed tests that fail with newer libxml (Thanks to knagano)

0.25_01 2007.04.27

    * Fixed XML::Atom::Base element accessor to work with attributes
      (Patch from LTjake and Jshirley for OpenSearch extension)
    * Make XML::Atom::Link easily subclassable
      (Patch from Simon Wistow for Google Calendar support)

0.25  2006.11.30
    * Fixed memory leaks in XML::Atom::Client
      (Reported by Brian Cassidy)

0.24  2006.11.25
    * Updated document to mention that ID creation is user's responsibility
    * Added $feed->as_xml_utf8 which always returns UTF-8 bytes string, rather than UTF-8 flagged one.
      This addresses annoying issues with UTF-8 vs. latin-1 (Thanks to Rui Vilela #21191)
    * Better fix for the hateful default: prefix issue in libxml2; now we
      remove the default\d* prefix on any nodes and set the proper namespace
      URI.
    * Fix to the test case since 0x242 is now printable character in bleadperl
      (Thanks to Andreas Koenig and Steve Peters)
    * Skip tests if 'euc-jp' is unknown encoding on your XML library
      (via CPAN testers)

0.23  2006.08.27
    * Fixed the method to get xml:lang and xml:base due to the
      XML::LibXML 1.60 change which invalidated it.

0.22  2006.07.24
    * Refactored internal element accessors by eating the new dog food
      mk_elem_accessors and mk_object_list_accessor.
    * Added support of Atom 1.0 <category> parse and generation
    * Added $thing->links and $thing->categories as a moniker method
      that returns an array reference in a scalar context
    * Fixed a bug in $content->body() where it accidentally thinks the
      content is not a valid Unicode string even if it is, if you call
      eval {} in elsewhere in the code and $@ is left set.
      (Thanks to Chris Dent for the patch)

0.21  2006.07.13
    * propagate $entry's version when we create content element off of entry
      using $entry->content("foo") syntax.
      (Thanks to Simon Wistow for spotting this bug)

0.20  2006.07.12
    * Lots of refactoring, which simplifies much of the code in the various
      construct modules (Person, Link, etc), and which should also make it
      much more straightforward to add extension classes in the future.
    * Added global $XML::Atom::ForceUnicode flag to return everything as Unicode flagged
      (Suggested by many people)
    * Added global $XML::Atom::DefaultVersion flag to set default
      version number for generated Atom feed. Defaults to 0.3 (for backward compatibility)
    * Added support for atom:content @type in Atom 1.0 feeds
      (Suggested by many people, especially Chris Dent and Andy Lester from Socialtext)

0.19  2006.03.19
    * Fix 0.18 bug where renaming stuff was totally broken.

0.18  2006.03.16
    * Support Atom 0.3 -> 1.0 renaming bits (issued -> published,
      modified -> updated, tagline -> subtitle)
    * $atom->content->body doesn't return Unicode flagged variable
      anymore, even if it's text/ data. Now it just returns UTF-8 bytes.
      (Thanks to Garth Webb)

0.17  2006.02.22
    * Fixed problem with XML::XPath

0.16  2005.11.22
    * Fixed a terrible bug when you set binary data to $entry->content

0.15  2005.11.01
    - Fixed bug found in mode => 'insert'
      (Thanks to Dominic Mitchell)

0.14  2005.10.21
    - Fixed bug that it decodes binary data as UTF-8 octet
      (Thanks to Mahlon E. Smith)
    - Don't eat up STDIN in cgi-mode
      (Thanks to Bayle Shanks)

0.13_02
    - Now supports insert mode, by passing hash reference in add_entry
      $feed->add_entry($entry, { mode => 'insert' });
      (Thanks to Dominic Mitchell)

0.13_01  2005.09.13
    - Now supports creating version 1.0 feed by passing new(Version => 1.0)
    - Be more strict in utf-8 handling and base64ing
      (Thanks to Dave Rolsky)

0.13  2005.08.18
    - Bumped up the version

0.12_02
    - $feed->version now returns 1.0 when xmlns patches with that for 1.0
    - Fixed segmentation fault problem with <atom:content> longer than 2.5k
      (Thanks to Chris Dent and Ryan King)
    - No unicode decode hack on content withou mode="xml"

0.12_01  2005.07.19
    - Added Atom 1.0 feed support for parsing
    - Hacked Unicode entity in $content->body
    - Added $entry->contributor
    - $entry->contributor and $entry->person returns list in list context
    - Added $content->lang and $content->base (xml:lang and xml:base)
    - Make sure $feed->as_xml doesn't set utf-8 flag

0.12  2005.06.07
    - Documentation fixes for XML::Atom::Server. Thanks to Tatsuhiko
      Miyagawa for the patch.
    - Removed XML::LibXSLT usage. Too much pain for too little gain.
      To be clear: all it was doing was namespace normalization, so

lib/XML/Atom/Feed.pm
# $Id: Feed.pm 70 2006-09-16 06:39:51Z miyagawa $

package XML::Atom::Feed;
use strict;
use base qw( XML::Atom::Thing );

use XML::Atom;
use XML::Atom::Entry;
BEGIN {
    if (LIBXML) {
        *entries = \&entries_libxml;
        *add_entry = \&add_entry_libxml;
    } else {
        *entries = \&entries_xpath;
        *add_entry = \&add_entry_xpath;
    }
}

sub init {
    my $atom = shift;
    my %param = @_ == 1 ? (Stream => $_[0]) : @_;
    if (UNIVERSAL::isa($param{Stream}, 'URI')) {
        my @feeds = __PACKAGE__->find_feeds($param{Stream});
        return $atom->error("Can't find Atom file") unless @feeds;
        my $ua = LWP::UserAgent->new;
        my $req = HTTP::Request->new(GET => $feeds[0]);
        my $res = $ua->request($req);
        if ($res->is_success) {
            $param{Stream} = \$res->content;
        }
    }
    $atom->SUPER::init(%param);
}

sub find_feeds {
    my $class = shift;
    my($uri) = @_;
    my $ua = LWP::UserAgent->new;
    my $req = HTTP::Request->new(GET => $uri);
    my $res = $ua->request($req);
    return unless $res->is_success;
    my @feeds;
    if ($res->content_type eq 'text/html' || $res->content_type eq 'application/xhtml+xml') {
        my $base_uri = $uri;
        my $find_links = sub {
            my($tag, $attr) = @_;
            if ($tag eq 'link') {
                return unless $attr->{rel};
                my %rel = map { $_ => 1 } split /\s+/, lc($attr->{rel});
                (my $type = lc $attr->{type}) =~ s/^\s*//;
                $type =~ s/\s*$//;
                push @feeds, URI->new_abs($attr->{href}, $base_uri)->as_string
                   if $rel{alternate} &&
                      $type eq 'application/atom+xml';
            } elsif ($tag eq 'base') {
                $base_uri = $attr->{href};
            }
        };
        require HTML::Parser;
        my $p = HTML::Parser->new(api_version => 3,
                                  start_h => [ $find_links, "tagname, attr" ]);
        $p->parse($res->content);
    } else {
        @feeds = ($uri);
    }
    @feeds;
}

sub element_name { 'feed' }

sub language {
    my $feed = shift;
    if (LIBXML) {
        my $elem = $feed->elem;
        if (@_) {
            $elem->setAttributeNS('http://www.w3.org/XML/1998/namespace',
                'lang', $_[0]);
        }
        return $elem->getAttributeNS('http://www.w3.org/XML/1998/namespace', 'lang');
    } else {
        if (@_) {
            $feed->elem->setAttribute('xml:lang', $_[0]);
        }
        return $feed->elem->getAttribute('xml:lang');
    }
}

sub version {
    my $feed = shift;
    my $elem = $feed->elem;
    if (@_) {
        $elem->setAttribute('version', $_[0]);
    }
    $elem->getAttribute('version') || $feed->SUPER::version(@_);
}

sub entries_libxml {
    my $feed = shift;
    my @res = $feed->elem->getElementsByTagNameNS($feed->ns, 'entry') or return;
    my @entries;
    for my $res (@res) {
        my $entry = XML::Atom::Entry->new(Elem => $res->cloneNode(1));
        push @entries, $entry;
    }
    @entries;
}

sub entries_xpath {
    my $feed = shift;
    my $set = $feed->elem->find("descendant-or-self::*[local-name()='entry' and namespace-uri()='" . $feed->ns . "']");
    my @entries;
    for my $elem ($set->get_nodelist) {
        ## Delete the link to the parent (feed) element, and append
        ## the default Atom namespace.
        $elem->del_parent_link;
        my $ns = XML::XPath::Node::Namespace->new('#default' => $feed->ns);
        $elem->appendNamespace($ns);
        my $entry = XML::Atom::Entry->new(Elem => $elem);
        push @entries, $entry;
    }
    @entries;
}

sub add_entry_libxml {
    my $feed = shift;
    my($entry, $opt) = @_;
    $opt ||= {};
    # When doing an insert, we try to insert before the first <entry> so
    # that we don't screw up any preamble.  If there are no existing
    # <entry>'s, then fall back to appending, which should be
    # semantically identical.
    my ($first_entry) =
        $feed->elem->getChildrenByTagNameNS($entry->ns, 'entry');
    if ($opt->{mode} && $opt->{mode} eq 'insert' && $first_entry) {
        $feed->elem->insertBefore($entry->elem, $first_entry);
    } else {
        $feed->elem->appendChild($entry->elem);
    }
}

sub add_entry_xpath {
    my $feed = shift;
    my($entry, $opt) = @_;
    $opt ||= {};
    my $set = $feed->elem->find("*[local-name()='entry' and namespace-uri()='" . $entry->ns . "']");
    my $first_entry = $set ? ($set->get_nodelist)[0] : undef;
    if ($opt->{mode} && $opt->{mode} eq 'insert' && $first_entry) {
        $feed->elem->insertBefore($entry->elem, $first_entry);
    } else {
        $feed->elem->appendChild($entry->elem);
    }
}

__PACKAGE__->mk_elem_accessors(qw( generator ));

__PACKAGE__->_rename_elements('modified' => 'updated');
__PACKAGE__->_rename_elements('tagline' => 'subtitle');

1;
__END__

=head1 NAME

XML::Atom::Feed - Atom feed

=head1 SYNOPSIS

    use XML::Atom::Feed;
    use XML::Atom::Entry;
    my $feed = XML::Atom::Feed->new;
    $feed->title('My Weblog');
    $feed->id('tag:example.com,2006:feed-id');
    my $entry = XML::Atom::Entry->new;
    $entry->title('First Post');
    $entry->id('tag:example.com,2006:entry-id');
    $entry->content('Post Body');
    $feed->add_entry($entry);
    $feed->add_entry($entry, { mode => 'insert' });

    my @entries = $feed->entries;
    my $xml = $feed->as_xml;

    ## Get a list of the <link rel="..." /> tags in the feed.
    my $links = $feed->link;

    ## Find all of the Atom feeds on a given page, using auto-discovery.
    my @uris = XML::Atom::Feed->find_feeds('http://www.example.com/');

    ## Use auto-discovery to load the first Atom feed on a given page.
    my $feed = XML::Atom::Feed->new(URI->new('http://www.example.com/'));

=head1 USAGE

=head2 XML::Atom::Feed->new([ $stream ])

Creates a new feed object, and if I<$stream> is supplied, fills it with the
data specified by I<$stream>.

Automatically handles autodiscovery if I<$stream> is a URI (see below).

Returns the new I<XML::Atom::Feed> object. On failure, returns C<undef>.

I<$stream> can be any one of the following:

=over 4

=item * Reference to a scalar

This is treated as the XML body of the feed.

=item * Scalar

This is treated as the name of a file containing the feed XML.

=item * Filehandle

This is treated as an open filehandle from which the feed XML can be read.

=item * URI object

This is treated as a URI, and the feed XML will be retrieved from the URI.

If the content type returned from fetching the content at URI is
I<text/html>, this method will automatically try to perform auto-discovery
by looking for a I<E<lt>linkE<gt>> tag describing the feed URL. If such
a URL is found, the feed XML will be automatically retrieved.

If the URI is already of a feed, no auto-discovery is necessary, and the
feed XML will be retrieved and parsed as normal.

=back

=head2 XML::Atom::Feed->find_feeds($uri)

Given a URI I<$uri>, use auto-discovery to find all of the Atom feeds linked
from that page (using I<E<lt>linkE<gt>> tags).

Returns a list of feed URIs. 

=head2 $feed->link

If called in scalar context, returns an I<XML::Atom::Link> object
corresponding to the first I<E<lt>linkE<gt>> tag found in the feed.

If called in list context, returns a list of I<XML::Atom::Link> objects
corresponding to all of the I<E<lt>linkE<gt>> tags found in the feed.

=head2 $feed->add_link($link)

Adds the link I<$link>, which must be an I<XML::Atom::Link> object, to
the feed as a new I<E<lt>linkE<gt>> tag. For example:

    my $link = XML::Atom::Link->new;
    $link->type('text/html');
    $link->rel('alternate');
    $link->href('http://www.example.com/');
    $feed->add_link($link);

=head2 $feed->add_entry($entry)

Adds the entry I<$entry>, which must be an I<XML::Atom::Entry> object,
to the feed. If you want to add an entry before existent entries, you can pass optional hash reference containing C<mode> value set to C<insert>.

  $feed->add_entry($entry, { mode => 'insert' });

=head2 $feed->entries

Returns list of XML::Atom::Entry objects contained in the feed.

=head2 $feed->language

Returns the language of the feed, from I<xml:lang>.

=head2 $feed->author([ $author ])

Returns an I<XML::Atom::Person> object representing the author of the entry,
or C<undef> if there is no author information present.

If I<$author> is supplied, it should be an I<XML::Atom::Person> object
representing the author. For example:

    my $author = XML::Atom::Person->new;
    $author->name('Foo Bar');
    $author->email('foo@bar.com');
    $feed->author($author);

=head2 $feed->id([ $id ])

Returns an id for the feed. If I<$id> is supplied, set the id. When
generating the new feed, it is your responsibility to generate unique
ID for the feed and set to XML::Atom::Feed object. You can use I<http>
permalink, I<tag> URI scheme or I<urn:uuid> for handy.

=head1 UNICODE FLAGS

By default, XML::Atom takes off all the Unicode flag fro mthe feed content. For example,

  my $title = $feed->title;

the variable C<$title> contains UTF-8 bytes without Unicode flag set,
even if the feed title contains some multibyte chracters.

If you don't like this behaviour and wants to andle everything as
Unicode characters (rather than UTF-8 bytes), set
C<$XML::Atom::ForceUnicode> flag to 1.

  $XML::Atom::ForceUnicode = 1;

then all the data returned from XML::Atom::Feed object and
XML::Atom::Entry object etc., will have Unicode flag set.

The only exception will be C<< $entry->content->body >>, if content
type is not text/* (e.g. image/gif). In that case, the content body is
still binary data, without Unicode flag set.

=head1 CREATING ATOM 1.0 FEEDS

By default, XML::Atom::Feed and other classes (Entry, Link and
Content) will create entities using Atom 0.3 namespaces. In order to
create 1.0 feed and entry elements, you can set I<Version> as a
parameter, like:

  $feed = XML::Atom::Feed->new(Version => 1.0);
  $entry = XML::Atom::Entry->new(Version => 1.0);

Setting those Version to every element would be sometimes painful. In
that case, you can override the default version number by setting
C<$XML::Atom::DefaultVersion> global variable to "1.0".

  use XML::Atom;

  $XML::Atom::DefaultVersion = "1.0";

  my $feed = XML::Atom::Feed->new;
  $feed->title("blah");

  my $entry = XML::Atom::Entry->new;
  $feed->add_entry($entry);

  $feed->version; # 1.0

=head1 AUTHOR & COPYRIGHT

Please see the I<XML::Atom> manpage for author, copyright, and license
information.

=cut
